skimage/feature/corner.py

```diff
@@ -954,36 +974,48 @@ def corner_peaks(image, min_distance=1, threshold_abs=None, threshold_rel=None,
            [2, 3],
            [2, 2]])
     >>> corner_peaks(response)
-    array([[2, 2]])
+    array([[3, 3]])
 
     """
     if threshold_rel is None:
         threshold_rel = 0.1
         warn("Until the version 0.16, threshold_rel was set to 0.1 by default."
              "Starting from version 0.16, the default value is set to None."
-             "Until version 0.18, a None value corresponds to a threshold value of 0.1."
-             "The default behavior will match skimage.feature.peak_local_max.",
-             category=FutureWarning, stacklevel=2)
-
-    peaks = peak_local_max(image, min_distance=min_distance,
-                           threshold_abs=threshold_abs,
-                           threshold_rel=threshold_rel,
-                           exclude_border=exclude_border,
-                           indices=False, num_peaks=num_peaks,
-                           footprint=footprint, labels=labels,
-                           num_peaks_per_label=num_peaks_per_label)
-    if min_distance > 0:
-        coords = np.transpose(peaks.nonzero())
-        for r, c in coords:
-            if peaks[r, c]:
-                peaks[max((r - min_distance), 0):r + min_distance + 1,
-                      max((c - min_distance), 0):c + min_distance + 1] = False
-                peaks[r, c] = True
+             "Until version 0.18, a None value corresponds to a threshold "
+             "value of 0.1. The default behavior will match "
+             "skimage.feature.peak_local_max.", category=FutureWarning,
+             stacklevel=2)
+
+    # Get the coordinates of the detected peaks
+    coords = peak_local_max(image, min_distance=min_distance,
+                            threshold_abs=threshold_abs,
+                            threshold_rel=threshold_rel,
+                            exclude_border=exclude_border,
+                            indices=True, num_peaks=num_peaks,
+                            footprint=footprint, labels=labels,
+                            num_peaks_per_label=num_peaks_per_label)
+
+    if len(coords):
+        # Use KDtree to find the peaks that are too close to each others
+        tree = spatial.cKDTree(coords)
+
+        rejected_peaks = set()
+        for idx, point in enumerate(coords):
+            if idx not in rejected_peaks:
+                candidates = tree.query_ball_point(point, r=min_distance, p=p)
+                candidates.remove(idx)
+                rejected_peaks.update(candidates)
+
+        # Remove the peaks that are too close to each others
+        coords = np.delete(coords, tuple(rejected_peaks), axis=0)[::-1]
```

Comment:
Why are you inverting the coordinates here? Are they originally in decreasing order of height, btw?

Reply:
They are originally in decreasing order (which is good), but that reordering is necessary for tests backward compatibility :-(